solutions/platinum/usaco-817.mdx

```diff
@@ -0,0 +1,148 @@
+---
+id: usaco-817
+source: USACO Platinum 2018 February
+title: New Barns
+author: Justin Ji
+---
+
+<Spoiler title = "Hint 1">
+
+The furthest distance from a node is always to one endpoint of a diameter of
+the tree. After adding a node, how can we maintain the diameter of its tree?
+
+</Spoiler>
+
+<Spoiler title = "Answer to Hint 1">
+
+Consider using the diameter before adding this node to find the diameter after
+adding the node in question.
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+Consider answering all the queries online. As given in Hint 1, the furthest
+distance from a given node is to an endpoint of a diameter of our current tree.
+Now the problem boils down to keeping track of the diameter of each tree inside
+our graph after we add a node.
+
+Observe that if we add a leaf to the tree that becomes an endpoint of our
+new diameter, then the other endpoint of our new diameter is one of the endpoints
+of our old diameter. To understand why this is the case, consider laying out
+the tree in the following manner:
+
+![Example Graph](usaco-817/diameter.png)
+
+Say we add a child to node 11. We now have a new diameter starting from that
+child, and need to find a suitable endpoint. Note that it is always optimal
+to pick one endpoint of the current diameter because if there was a better
+option within one of the "subtrees", it would be the endpoint of the previous
+diameter.
+
+With this fact, we can easily maintain the diameter for each tree in our graph
+online. Finding the distance between each pair of nodes can be done using binary
+lifting.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+class Tree {
+  private:
+	const int q, log2dist;
+	vector<vector<int>> up;
+	vector<pair<int, int>> diameter;
+	vector<int> root;
+	vector<int> dep;
+	int curr = -1;
+
+  public:
+	Tree(int q)
+	    : q(q), log2dist(1 + (int)log2(q)), up(log2dist, vector<int>(q)),
```

Comment:
ceil instead?

Reply:
eh i personally would rather use as little floating point as possible

Reply:
i mean unless you can come up with an example where ceil doesn't work
you already have bits/stdc++ and std
and it gets the idea across much better

Reply:
eh ok sure, i still dont like it but I can kind of see why